File: 1915-check-if-one-string-swap-can-make-strings-equal/check-if-one-string-swap-can-make-strings-equal.py
class Solution(object):
    def areAlmostEqual(self, s1, s2):
        """
        :type s1: str
        :type s2: str
        :rtype: bool
        """
        if len(s1) != len(s2):  # Length mismatch
            return False  

        first_index, second_index = -1, -1
        mismatch_count = 0

        for i in range(len(s1)):
            if s1[i] != s2[i]:
                mismatch_count += 1

                if first_index == -1:
                    first_index = i
                elif second_index == -1:
                    second_index = i
                else:
                    return False  # More than two mismatches

        # If no mismatch, strings are already equal
        if mismatch_count == 0:
            return True

        # If only one mismatch, it's impossible to swap -> return False
        if mismatch_count != 2:
            return False

        # Swap characters and check if they become equal
        s1_list = list(s1)  
        s1_list[first_index], s1_list[second_index] = s1_list[second_index], s1_list[first_index]

        return "".join(s1_list) == s2
        

File: 1915-check-if-one-string-swap-can-make-strings-equal/check-if-one-string-swap-can-make-strings-equal.cpp
class Solution {
public:
    bool areAlmostEqual(string s1, string s2) {
        int n_s1 = s1.length();
        int n_s2 = s2.length();

        if (n_s1 != n_s2) return false; // Length mismatch

        int first_index = -1, second_index = -1;
        int mismatch_count = 0;

        for (int i = 0; i < n_s1; i++) {
            if (s1[i] != s2[i]) {
                mismatch_count++;

                if (first_index == -1) 
                    first_index = i;
                else if (second_index == -1) 
                    second_index = i;
                else 
                    return false; // More than two mismatches
            }
        }

        // If no mismatch, strings are already equal
        if (mismatch_count == 0) return true;

        // If only one mismatch, it's impossible to swap -> return false
        if (mismatch_count != 2) return false;

        // Swap and check if they become equal
        swap(s1[first_index], s1[second_index]);

        return s1 == s2;
    }
};
